qa/L0_data_compression/test.sh

```diff
@@ -70,7 +70,37 @@ fi
 set -e
 
 # End-to-end testing with simple model
+function run_data_compression_infer_client() {
+    local client_path=$1
+    local request_algorithm=$2
+    local response_algorithm=$3
+    local log_path=$4
+
+    local python_or_cpp=`echo -n "$client_path" | tail -c 3`
```

Comment:
[Optional] I'm only a little weary of using a hard count from the end to determine the file extension. Technically, this would work for all extensions I can imagine us using in the future e.g., `.py`, `.cc`, `.rs`, etc. However, if the extension was `.cpp` as the variable name suggests, this line would not store the intended string.

Maybe it would be more future-proof if we did something like:
```bash
filename=$(basename ${client_path})
extension=".${filename##*.}"
```

Reply:
This is a good suggestion if we are sure there is an extension (e.g. `file.py`, `file.cpp`), but there is no extension for the c++ executable, i.e.
```
$ path="../clients/simple_http_infer_client" && filename=$(basename ${path}) && extension=".${filename##*.}" && echo $extension
.simple_http_infer_client
```
which I think `.simple_http_infer_client` is no better than the last three characters `ent`.

I don't know if one day there might be a Java executable, e.g. `../clients/simple_http_java_infer_client`, and uses a different set of command line options than the c++ one, which we will have to come back and rewrite the test anyway.

What do you think about leaving it simple for now (just the last three characters)?

Reply:
Oof 🤦‍♂️ that's a really good point, I was stuck in source file land. In that case I am okay leaving it checking the last three characters!

Reply:
Personally I'm fine with current implementation, or just making it more straightforward and just defining what type each client is:
(may not be syntactically correct)
```bash
function run_data_compression_infer_client {
    ...
    if [ "$request_algorithm" != "" ] && [ "$response_algorithm" == "" ]; then
        $infer_client -v $request_algorithm_parameter >> $log_path 2>&1
        return $?
    fi
    if [ "$request_algorithm" == "" ] && [ "$response_algorithm" != "" ]; then
        $infer_client -v $response_algorithm_parameter >> $log_path 2>&1
        return $?
    fi

    $infer_client -v $request_algorithm_parameter $response_algorithm_parameter >> $log_path 2>&1
    if [ $? -ne 0 ]; then
       RET=1
       echo "<this specific test failed>"
    fi
}


function run_python_data_compression_test {
        local infer_client="python $client_path"
        local request_algorithm_parameter="--request-compression-algorithm $request_algorithm"
        local response_algorithm_parameter="--response-compression-algorithm $response_algorithm"
        run_data_compression_infer_client "${infer_client}" "$REQUEST_ALGORITHM" "$RESPONSE_ALGORITHM" "$CLIENT_LOG"
}

function run_cpp_data_compression_test {
  ...
}

python_clients="${SIMPLE_INFER_CLIENT_PY} ${SIMPLE_AIO_INFER_CLIENT_PY}"
cpp_clients = "${SIMPLE_INFER_CLIENT}"

    for REQUEST_ALGORITHM in "deflate" "gzip" ""; do
        for RESPONSE_ALGORITHM in "deflate" "gzip" ""; do
            if [ "$REQUEST_ALGORITHM" == "$RESPONSE_ALGORITHM" ]; then
                continue
            fi

            set +e
            for client in ${PYTHON_CLIENTS}; do
              run_python_data_compression_test
            }

            for client in ${CPP_CLIENTS}; do
              run_cpp_data_compression_test
            }
            set -e
        done
    done
```

but either way I don't think this part needs to be over-engineered